kbase/execution_engine2
# Fix timezone related bug / test failures

# Description of PR purpose/changes

Parts of the code were using naive datetimes for epoch time processing. This
meant that when run in an environment where the timezone was UTC (like the
docker container build from the provided dockerfile) everything worked.
However, when run in an environment with a different timezone (like my
laptop, PST) the tests would fail.

There were two problems. Firstly, the test was creating a naive timestamp
using `utcnow()` instead of `now()`, which causes the underlying timestamp to
differ when regenerated from the datetime object:
```
n [1]: from datetime import datetime, timezone

In [2]: utcnow = datetime.utcnow()

In [3]: now = datetime.now(tz=timezone.utc)

In [4]: print(utcnow, utcnow.timestamp())
2021-02-11 01:53:58.919172 1613037238.919172

In [5]: print(now, now.timestamp())
2021-02-11 01:54:01.017436+00:00 1613008441.017436
```
Note that `now` is timezone aware, while `utcnow` is not.

Secondly, timestamps were being parsed without timezone awareness in the
`_get_dummy_dates()` method, which caused the MongoDB `ObjectID`s to be
generated from the datetime incorrectly:

```
In [24]: ts = datetime.fromtimestamp(100000)

In [25]: ts
Out[25]: datetime.datetime(1970, 1, 1, 19, 46, 40)

In [26]: ObjectId.from_datetime(ts)
Out[26]: ObjectId('000116200000000000000000')

In [27]: ts = datetime.fromtimestamp(100000, tz=timezone.utc)

In [28]: ts
Out[28]: datetime.datetime(1970, 1, 2, 3, 46, 40, tzinfo=datetime.timezone.utc)

In [29]: ObjectId.from_datetime(ts)
Out[29]: ObjectId('000186a00000000000000000')
```

This commit also adds a Pipfile for ease of use of dependency managers that
take that format.

# Jira Ticket / Github Issue #
- [n/a] Added the Jira Ticket to the title of the PR e.g. (DATAUP-69 Adds a PR template)

# Testing Instructions
* Details for how to test the PR: 
- [X] Changes available by spinning up a local test suite
If you want to actually check these tests work, you must run them on a machine that is *not* on UTC time. I know they fail on PST time, but I'm not sure how far you have to get from UTC for the tests to fail.

# Dev Checklist:

- [n/a] I have commented my code, particularly in hard-to-understand areas
- [n/a] I have made corresponding changes to the documentation
- [?] My changes generate no new warnings - over 3500 warnings, too hard to check
- [n/a] I have added tests that prove my fix is effective or that my feature works - I can't, since the tests run in UTC time in Docker and on GHA
- [n/a] Any dependent changes have been merged and published in downstream modules

# Updating Version and Release Notes (if applicable)

- [n/a] [Version has been bumped](https://semver.org/) for each release
- [n/a] [Release notes](/RELEASE_NOTES.md) have been updated for each release (and during the merge of feature branches)
